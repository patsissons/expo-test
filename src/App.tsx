import React from 'react';
import { createStore, applyMiddleware } from 'redux';
import { Provider } from 'react-redux';
import { createEpicMiddleware } from 'redux-observable';
import { composeWithDevTools } from 'redux-devtools-extension'
import '../types';
import { reducers } from './reducers';
import { epics } from './epics';
import { AppNavigator, reactNavigationMiddleware } from './navigation/AppNavigator';

const store = createStore(
  reducers,
  composeWithDevTools(
    applyMiddleware(
      createEpicMiddleware(epics as any),
      reactNavigationMiddleware,
    ),
  ),
);

export class App extends React.Component {
  render() {
    return (
      <Provider store={ store }>
        <AppNavigator />
      </Provider>
    );
  }
}
