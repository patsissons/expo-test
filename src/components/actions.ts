import { Dispatch, bindActionCreators, Action } from 'redux';
import { NavigationParams } from 'react-navigation';
import { connect, Connect } from 'react-redux';
import { getTypeNames, createTypeNameMap } from '../utils';
import { Store } from './state';

export type Actions = Readonly<typeof actions>;

export interface ActionProps {
  readonly actions?: Actions;
}

export interface RequestAction<T = {}> extends Action {
  payload?: T;
}

export interface DemoPayload {
  amount: number;
}

function getRequestAction<T>(type: string, payload?: T): RequestAction<T> {
  return {
    type,
    payload,
  };
}

enum Navigation {
  NavigationBack,
  NavigationToMain,
  NavigationToLogin,
}

enum Demo {
  DemoIncrement,
  DemoDecrement,
}

export const actionTypes = {
  navigation: createTypeNameMap(Navigation),
  demo: createTypeNameMap(Demo),
};

const actions = {
  navigation: {
    goBack: (params?: NavigationParams) => getRequestAction(actionTypes.navigation.NavigationBack, params),
    toMain: (params?: NavigationParams) => getRequestAction(actionTypes.navigation.NavigationToMain, params),
    toLogin: (params?: NavigationParams) => getRequestAction(actionTypes.navigation.NavigationToLogin, params),
  },
  demo: {
    increment: (payload?: DemoPayload) => getRequestAction(actionTypes.demo.DemoIncrement, payload),
    decrement: (payload?: DemoPayload) => getRequestAction(actionTypes.demo.DemoDecrement, payload),
  },
};

function mapDispatchToActions(dispatch: Dispatch<Store>): ActionProps {
  return {
    actions: getTypeNames(actions)
      .reduce(
        (map, actionName) => {
          map[actionName] = bindActionCreators(actions[actionName], dispatch);

          return map;
        },
        {} as any,
      ),
  };
}

function connectToActionsHelper(...args: Array<any>) {
  args[1] = args[1] || mapDispatchToActions;

  return connect.apply(undefined, args);
}

export const connectToActions: Connect = connectToActionsHelper;
