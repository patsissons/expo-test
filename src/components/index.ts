export * from './state';
export * from './actions';
export * from './context';
export * from './AppComponent';
export * from './startup/Startup';
