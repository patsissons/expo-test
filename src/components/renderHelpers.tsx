import * as React from 'react';
import { Text } from 'react-native';
import { isString, isObject } from '../utils';

export function coerceToText(content: any) {
  if (isString(content)) {
    return (
      <Text>{ content }</Text>
    );
  }

  if (isObject(content)) {
    return (
      <Text { ...content } />
    );
  }

  if (React.isValidElement(content)) {
    return content;
  }

  return false;
}
