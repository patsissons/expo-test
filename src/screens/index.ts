export * from './MainScreen';
export * from './LoginScreen';
