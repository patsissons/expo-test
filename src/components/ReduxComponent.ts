import React from 'react';
import { StyleSheet } from 'react-native';
import { NavigationScreenProps } from 'react-navigation';
import { ActionProps, connectToActions } from './actions';

export namespace ReduxComponent {
  export interface StateProps<T> {
    state: T;
  }

  export interface Props<T> extends StateProps<T>, ActionProps {
  }

  export interface ScreenProps<T> extends Props<T>, NavigationScreenProps {
  }
}

export abstract class ReduxComponent<P extends ReduxComponent.Props<RS>, RS, S = {}, SS = any> extends React.Component<P, S, SS> {
  public static readonly connect = connectToActions;

  public static readonly defaultStyles = StyleSheet.create({
    element: {
    },
    container: {
      flex: 1,
      backgroundColor: '#fff',
      alignItems: 'center',
      justifyContent: 'center',
    },
  });

  public get defaultStyles() {
    return ReduxComponent.defaultStyles;
  }

  public get actions() {
    return this.props.actions!;
  }
}
