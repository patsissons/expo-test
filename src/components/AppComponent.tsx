import React from 'react';
import { NavigationScreenProps } from 'react-navigation';
import { ReduxProvider } from './context';
import * as styles from './styles';
import * as renderHelpers from './renderHelpers';
import { ContextState } from './state';

export namespace AppComponent {
  export interface Props<TState> extends ReduxProvider.ComponentProps<TState> {
  }

  export interface ScreenProps<TState, P = {}, O = {}> extends Props<TState>, NavigationScreenProps<P, O> {
  }

  export interface ContextProps {
    context: ContextState;
  }

  export interface ContextNavigationScreenProps extends NavigationScreenProps {
    screenProps: ContextProps;
  }
}

export abstract class AppComponent<P extends AppComponent.Props<TState>, TState, S = {}, SS = any> extends React.Component<P, S, SS> {
  constructor(props: P) {
    super(props);

    this.state = {} as S;
  }

  protected get L() {
    return this.props.context.localization;
  }

  protected get R() {
    return renderHelpers;
  }

  protected get S() {
    return styles;
  }
}
