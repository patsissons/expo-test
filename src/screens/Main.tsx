import React from 'react';
import { View, Button } from 'react-native';
import { MainState, AppComponent, withState } from '../components';
import { Demo } from '../components/demo/Demo';

class MainScreen extends AppComponent<Main.ScreenProps, MainState> {
  public static navigationOptions(params: AppComponent.ContextNavigationScreenProps) {
    return {
      title: params.screenProps.context.localization.MainScreen,
    };
  }

  render() {
    return (
      <View style={ this.S.common.container }>
        <Button title={ this.L.Login } onPress={ this.handleLogin.bind(this) }>{ this.L.Login }</Button>
        <Demo />
      </View>
    );
  }

  protected handleLogin() {
    this.props.actions.navigation.toLogin();
  }
}

export namespace Main {
  export interface Props {
  }

  export interface ScreenProps extends AppComponent.ScreenProps<MainState>, Props {
  }
}

export const Main = withState(
  MainScreen,
  x => x.main,
);
