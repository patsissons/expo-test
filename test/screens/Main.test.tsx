import React from 'react';
import { shallow } from 'enzyme';
import { Button } from 'react-native';
import configureStore from 'redux-mock-store';
import { Store, MainState, actionTypes } from '../../src/components';
import { Main } from '../../src/screens/Main';

const defaultState: MainState = {
  demo: {
    count: 0,
  },
};

function createStoreProps(state?: MainState) {
  state = state || defaultState;

  const mockStore = configureStore<Partial<Store>>();

  return { store: mockStore({ main: state }) };
}

describe('screens', () => {
  describe('Main', () => {
    it('renders without crashing', () => {
      // Arrange
      const storeProps = createStoreProps();
      const component = (<Main { ...storeProps as any } />);

      // Act
      const result = shallow(component).dive();

      // Assert
      expect(result).toMatchSnapshot();
    });

    it('navigates to login after pressing the button', () => {
      // Arrange
      const storeProps = createStoreProps();
      const component = (<Main { ...storeProps as any } />);

      // Act
      const result = shallow(component).dive();
      result.find(Button).simulate('press');

      // Assert
      expect(storeProps.store.getActions()).toEqual([{ type: actionTypes.navigation.NavigationToLogin }]);
    });
  });
});
