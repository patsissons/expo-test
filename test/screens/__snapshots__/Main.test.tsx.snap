// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`screens Main renders without crashing 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MainScreen
    actions={
      Object {
        "context": Object {
          "loadEnv": [Function],
          "loadLocale": [Function],
        },
        "demo": Object {
          "decrement": [Function],
          "increment": [Function],
        },
        "navigation": Object {
          "goBack": [Function],
          "toLogin": [Function],
          "toMain": [Function],
        },
      }
    }
    context={
      Object {
        "env": Object {
          "RELEASE": "local",
        },
        "locale": "en",
        "localization": Object {
          "CurrentStateIs": "Current state is",
          "Increment": "Increment",
          "IncrementCount": "Increment Count",
          "Login": "Login",
          "LoginScreen": "Login Screen",
          "MainScreen": "Main Screen",
        },
      }
    }
    state={
      Object {
        "demo": Object {
          "count": 0,
        },
      }
    }
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    storeSubscription={
      Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <Button
          onPress={[Function]}
          title="Login"
        >
          Login
        </Button>,
        <Connect(DemoComponent) />,
      ],
      "style": Object {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "flex": 1,
        "justifyContent": "center",
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "Login",
          "onPress": [Function],
          "title": "Login",
        },
        "ref": null,
        "rendered": "Login",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <Button
            onPress={[Function]}
            title="Login"
          >
            Login
          </Button>,
          <Connect(DemoComponent) />,
        ],
        "style": Object {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "flex": 1,
          "justifyContent": "center",
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Login",
            "onPress": [Function],
            "title": "Login",
          },
          "ref": null,
          "rendered": "Login",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
